### ISSUE TITLE
Data source: pruning doesn't work

### ISSUE THREAD

**Issue by dpedu** [grafana/grafana#94645]
tl;dr the `prune:true` method for removing provisioned data sources doesn't seem to work.

### What happened?

I added a VictoriaMetrics data source using the provisioning file that the project provides, seen here: https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/deployment/docker/provisioning/datasources/victoriametrics-datasource/cluster.yml

Later, since I added this to familiarize myself with the provisioning process, I wanted to remove it.

First, as documented here - https://grafana.com/docs/grafana/latest/administration/provisioning/ - I tried removing the existing provisioning file and adding a new file in `/etc/grafana/provisioning/datasources/delete.yml` with the contents:

```
apiVersion: 1
prune: true
```

I'm using the official image, so to check if it is in the right place within the image:

```
sudo docker exec -it mygrafana bash
9058f458b11b:/usr/share/grafana$ cat /etc/grafana/provisioning/datasources/delete.yml
apiVersion: 1
prune: true
```

But, when I restarted Grafana the data source was not deleted.

To double-check that I had indeed placed the file in the right place, I decided to add a second VictoriaMetrics datasource with a different name through the provisioning system. It was added successfully. This was via another yml file in the same dir as  the one mentioned above.

Since the pruning method wasn't working, I removed the file and I tried the other deletion method mentioned in the docs above. Now my `delete.yml`, the only file in the `provisioning/datasources` dir, looked like:

```
apiVersion: 1
prune: true

deleteDatasources:
  - name: VictoriaMetrics
    orgId: 1
```

And this time, it did indeed delete the datasource I named:

```
logger=provisioning.datasources t=2024-10-11T18:49:48.26920848Z level=info msg="deleted datasource based on configuration" name=VictoriaMetrics
```

However, the second VictoriaMetrics was not removed. I believe it should have been given the `prune: true` option.

### What did you expect to happen?

I expected the data source be deleted since there is no longer a config file for it and the `prune: true` option is set.

### Did this work before?

I have never tried this before.

### How do we reproduce it?

I have replicated this outside of the environment I encountered it in, starting from a blank slate. 

* First, set up directories for running dockerized grafana:

```
mkdir grafanatest
cd grafanatest/
mkdir data
mkdir provisioning
cd data/
mkdir plugins
cd plugins/
wget -q  https://github.com/VictoriaMetrics/victoriametrics-datasource/releases/download/v0.9.1/victoriametrics-datasource-v0.9.1.tar.gz
tar xf victoriametrics-datasource-v0.9.1.tar.gz
cd ../../
cd provisioning/
mkdir datasources
cd datasources/
wget https://raw.githubusercontent.com/VictoriaMetrics/VictoriaMetrics/22d3f679085d246202f26990eb747f2b7370e2c8/deployment/docker/provisioning/datasources/victoriametrics-datasource/cluster.yml
```

With pwd grafanatest, the following items are present:

```
./provisioning/datasources/cluster.yml
./data/plugins/victoriametrics-datasource/... (lots of files from the plugin)
```

* Start grafana, using the following command:

```
docker run \
    --rm \
    --name=mygrafana \
    -e "GF_SECURITY_ADMIN_PASSWORD=admin" \
    -e "GF_SERVER_HTTP_PORT=8080" \
    -e "GF_PLUGINS_ALLOW_LOADING_UNSIGNED_PLUGINS=victoriametrics-datasource" \
    -v /Users/dave/Downloads/grafanatest/data:/var/lib/grafana \
    -v /Users/dave/Downloads/grafanatest/provisioning:/etc/grafana/provisioning \
    -p 8080:8080 \
    grafana/grafana:11.2.2
```

* Log into the UI, and verified that the plugin is enabled and the datasource is present.

* Stop grafana.

* Delete cluster.yml from above

* Start grafana again. Log into the UI, and verified that the datasource is still present, to prove it was persisted to the database and that the database persists

* Stop grafana

* Create `./provisioning/datasources/delete.yml`, with the contents:

```
apiVersion: 1
prune: true
```

* Start grafana 

* Log into the UI and see that the datasource is still there. It should have been pruned. 

With `GF_LOG_LEVEL=debug`, I do not see my `delete.yml` ever mentioned.

### Is the bug inside a dashboard panel?

No

### Environment (with versions)?

Grafana: 11.2.2
OS: Docker for macOs
Browser: Chrome

### Grafana platform?

Docker

### Datasource(s)?

victoriametrics-datasource

**Comment by andresmgot** [comment]
Hello @dpedu, the documentation may be a bit confusing here. When it says that `prune: true` should be added to the "root", it means the root level of the yaml file used to provision the datasource. You can see an example [here](https://grafana.com/docs/grafana/latest/administration/provisioning/#example-data-source-configuration-file), or this simplified version:

```yaml
apiVersion: 1
prune: true
datasources:
  - name: test_graphite_with_prune
    type: graphite
    access: proxy
    url: http://localhost:8080
```

With the provisioning file above, after you delete the file, the datasource `test_graphite_with_prune` will be automatically deleted the next time Grafana is started. I have tried to reword the documentation a bit to make it more clear: https://github.com/grafana/grafana/pull/95059

Let me close this issue, but if you still find trouble using this feature, feel free to re-open it.

**Comment by dpedu** [comment]
Hi @andresmgot, and thank you but I am still having trouble unfortunately. 

I did have the `prune: true` in a top-level key, or the root of my provisioning yaml file. However, I had renamed the file from cluster.yml to delete.yml in my original example.

However, if I do not rename the file it seems to result in the same problem. 

I'm using cluster.yml from the link above to provision the data source. I next tried changing the contents of this file instead, adding `prune: true` as you described. I also changed `datasources` to an empty list:

```
apiVersion: 1
prune: true
datasources: []
```

But the datasource wasn't deleted. 

I played around with this a bit more, and I think I've found a better clue:

- If I add `prune: true` to the file, restart grafana, THEN remove the datasource, and restart grafana again, it is removed.
- If I add `prune: true` AND remove the datasource and THEN restart grafana, it is _not_ removed.

So, I think the missing detail here is: Grafana needs to be restarted for the `prune: true` option to be picked up. In my opinion, this is a bug.

**Comment by andresmgot** [comment]
> * If I add `prune: true` to the file, restart grafana, THEN remove the datasource from the file, and restart grafana again, it is removed.

Yes, this is how it works.

> Grafana needs to be restarted for the prune: true option to be picked up

That's correct and expected. Note that the provisioning process happens only when starting Grafana. Any change in provisioning files requires to restart Grafana in order to be evaluated. In this case, this is what happens:

 - When the provisioning process reads a file with `prune: true`, it saves in the database the datasources info and marks them with `is_prunable: true`.
 - When the provisioning process is run again (after a restart), if there is a datasource in the database with `is_prunable: true` that is no longer part of any provisioning file, then it deletes it.

Hope that clarifies the behavior.
